libraries/Microsoft.Bot.Builder.Dialogs/Memory/DialogStateManager.cs

```diff
@@ -1,701 +0,0 @@
-﻿// Copyright (c) Microsoft Corporation. All rights reserved.
-// Licensed under the MIT License.
-
-using System;
-using System.Collections;
-using System.Collections.Generic;
-using System.Diagnostics;
-using System.Globalization;
-using System.Linq;
-using System.Threading;
-using System.Threading.Tasks;
-using Microsoft.Bot.Builder.Dialogs.Memory.Scopes;
-using Newtonsoft.Json.Linq;
-
-namespace Microsoft.Bot.Builder.Dialogs.Memory
-{
-    /// <summary>
-    /// The DialogStateManager manages memory scopes and pathresolvers
-    /// MemoryScopes are named root level objects, which can exist either in the dialogcontext or off of turn state
-    /// PathResolvers allow for shortcut behavior for mapping things like $foo -> dialog.foo.
-    /// </summary>
-#pragma warning disable CA1710 // Identifiers should have correct suffix (We can't rename this class without breaking binary compat)
-    public class DialogStateManager : IDictionary<string, object>
```

Comment:
From reviewing the usage of Dialog.State property, I am inclined to preserver some functionality of DialogStateManager rather than completely removing it. I would suggest we:

1. Change `public class DialogStateManager : ConcurrentDictionary<string, object>` and move it to another namespace
2. Remove all the inner code and just add only a few generic methods to be able to retrieve strong typed versions of the dictionary (making sure we follow the default behavior for dictionaries).

For example, if we update the `DialogStateManager` class to be: 

``` csharp
public class DialogStateManager : ConcurrentDictionary<string, object>
{
    public T GetValue<T>(string key)
    {
        return (T)this[key];
    }

    public bool TryGetValue<T>(string key, out T value)
    {
        if (TryGetValue(key, out var dictionaryValue))
        {
            value = (T)dictionaryValue;
            return true;
        }

        value = default;
        return false;
    }
}
```

We can write some unit tests like this to try it out (we may need to do serialization and exception tests too):

``` csharp
public class DialogStateManagerTests
{
    private readonly List<string> _someList;
    private readonly DialogStateManager _sut;

    public DialogStateManagerTests()
    {
        _someList = new List<string>
        {
            "first",
            "second"
        };
        _sut = new DialogStateManager
        {
            ["someInt"] = 42,
            ["someString"] = "hello",
            ["someBool"] = true,
            ["someObject"] = _someList
        };
    }

    [Fact]
    public void GetValue()
    {
        Assert.Equal(42, _sut["someInt"]);
        Assert.Equal(42, _sut.GetValue<int>("someInt"));

        Assert.Equal("hello", _sut["someString"]);
        Assert.Equal("hello", _sut.GetValue<string>("someString"));

        Assert.Equal(true, _sut["someBool"]);
        Assert.True(_sut.GetValue<bool>("someBool"));

        Assert.Equal(_someList, _sut["someObject"]);
        Assert.Equal(_someList, _sut.GetValue<List<string>>("someObject"));
    }

    [Theory]
    [InlineData("")]
    [InlineData("NotFound")]
    public void TryGetValue(string keySuffix)
    {
        Assert.Equal(_sut.TryGetValue<int>($"someInt{keySuffix}", out var intValue) ? 42 : 0, intValue);
        Assert.Equal(_sut.TryGetValue<string>($"someString{keySuffix}", out var stringValue) ? "hello" : null, stringValue);
        Assert.Equal(_sut.TryGetValue<bool>($"someBool{keySuffix}", out var boolValue) ? true : false, boolValue);
        Assert.Equal(_sut.TryGetValue<List<string>>($"someObject{keySuffix}", out var objectValue) ? _someList : null, objectValue);
    }
}
```

I think this will give developers a way of accessing state in a similar way as they do today and it provides helpers to access state and return string typed objects.

What do you think?

Reply:
My understanding of how DialogStateManager integrates with the sdk is primarily around MemoryScopes. I think the type coercion you're mentioning is demonstrated here: https://github.com/microsoft/botbuilder-dotnet/pull/6076

The issue is, as objects are pulled out of storage, they will not have type info. So, the generic methods you've outlined above will not work as expected.

Reply:
OK, let's revisit this one once we review the other PR about type handling.